File: CPP/25. Reverse Nodes in k-Group.cpp
#include <iostream>
#include "printFunctions.hpp"

using namespace std;

class Solution {
public:
	ListNode* reverseKGroup(ListNode* head, int k) {
		ListNode* curr = head;
		for (int i = 0; i < k;i++) {
			if (curr==NULL) {
				return head;
			}
			curr = curr->next;
		}

		ListNode* prev = nullptr;
		curr = head;

		for (int i = 0; i < k; i++) {
			ListNode* next = curr->next;
			curr->next = prev;
			prev = curr;
			curr = next;
		}

		head->next = reverseKGroup(curr, k);
		return prev;
	}
};

int main() {
	Solution sol;
	int k = 3;
	ListNode* head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
	ListNode* result = sol.reverseKGroup(head, k);
	printListNode(result);
	return 0;
}

File: Python/25. Reverse Nodes in k-Group.py
from re import T
from typing import Optional
from ListNode import *

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        curr = head
        for _ in range(k):
            if not curr:
                return head
            curr = curr.next

        prev = None
        curr = head
        for _ in range(k):
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
            
        head.next = self.reverseKGroup(curr, k)
        return prev

if __name__ == '__main__':
    k = 3
    head = ListNode(1,ListNode(2,ListNode(3,ListNode(4,ListNode(5)))))
    result = Solution().reverseKGroup(head,k)
    print(result)